Enable to load models with low opset version

sungchul2 opened the pull request:
### Changes

<!--- What was changed (briefly), how to reproduce (if applicable), what the reviewers should focus on -->
- Enable to load models with low opset version
- Add to change tensor shape in Accuracy

### Reason for changes

<!--- Why should the change be applied -->
1. The models with opset==7, 8, 9 have not only input tensor, but also every weight and bias in original_model.graph.input. But the model with opset==12, which have only input tensor in original_model.graph.input, can be only loaded.
<opset==7, 8, 9>
![image](https://user-images.githubusercontent.com/97221135/182327596-b99ecb66-6eb4-4764-84b1-6c1ff564c0d6.png)
<opset==12>
![image](https://user-images.githubusercontent.com/97221135/182327696-26046af0-ce3c-4b66-b229-3eaa9e97945c.png)

2. densenets output a 4d tensor , but Accuracy metric cannot use it because it can only handle a 2d tensor.

### Related tickets

<!--- Post the numerical ID of the ticket, if available -->
88636

### Tests

<!--- How was the correctness of changes tested and whether new tests were added -->

kshpv commented:
@sungchul2 thanks! Could you also add a test that will failed without your fix and pass with it?

kshpv commented on examples/experimental/onnx/classification/onnx_ptq_classification.py:
```diff
@@ -38,15 +39,7 @@ def run(onnx_model_path: str, output_model_path: str,
     nncf_logger.info("The model is loaded from {}".format(onnx_model_path))
 
-    if input_shape is None:
-        nncf_logger.info(
```
Please save this message

sungchul2 replied:
Done

kshpv commented on nncf/experimental/onnx/datasets/imagenet_dataset.py:
```diff
@@ -11,7 +11,7 @@
  limitations under the License.
 """
 
-from typing import Optional, Tuple
```
In NNCF styleguide https://github.com/openvinotoolkit/nncf/blob/develop/docs/styleguide/PyGuide.md we have the rule - Imports should usually be on separate lines

sungchul2 replied:
Yes, but typing module looks exceptional.
![image](https://user-images.githubusercontent.com/97221135/183320664-93bdbc8a-82f3-44b0-b372-e63e4bf91d6c.png)

kshpv commented on nncf/experimental/onnx/datasets/imagenet_dataset.py:
```diff
@@ -47,18 +47,30 @@ def __len__(self) -> int:
+                    input_shape = _input_shape
+                    input_keys = _input.name
+                else:
+                    if input_shape[2] < _input_shape[2]:
```
Why does it work like that? Could you add some comments in the code?

sungchul2 replied:
Done

vinnamkim commented on examples/experimental/onnx/classification/onnx_ptq_classification.py:
```diff
@@ -41,12 +41,8 @@ def run(onnx_model_path: str, output_model_path: str,
-    input_keys = [node.name for node in original_model.graph.input]
-    if len(input_keys) != 1:
-        raise RuntimeError(
-            f"The number of inputs should be 1(!={len(input_keys)}).")
+    input_shape, input_keys = infer_input_shape(original_model, input_shape)
```
I recommend you to solve this problem by adding an input argument of input tensor name rather than picking up the input tensor automatically. We can ask user to give the proper input tensor name if we fail to find the input tensor.

sungchul2 replied:
I agree. I fixed to ask user either the input tensor's name or shape and to find the proper tensor shape and name.

vinnamkim commented on nncf/experimental/post_training/api/metric.py:
```diff
@@ -134,8 +134,13 @@ def update(self, outputs: NNCFData, targets: NNCFData) -> None:
 
         if outputs.ndim != 2:
-            raise ValueError('The accuracy metric should be calculated on 2d outputs. '
-                             f'However, the outputs has ndim={outputs.ndim}.')
+            if outputs.ndim == 4 and (outputs.shape[2] == 1 and outputs.shape[3] == 1):
+                # densenet outputs (1,1000,1,1). It is able to change tensor shape.
+                outputs = outputs.squeeze(axis=(2, 3))
+            else:
+                raise ValueError('The accuracy metric should be calculated on 2d outputs. '
+                                 f'However, the outputs has ndim={outputs.ndim}.')
```
Please avoid nested if-else statements for the better readability.

```
if:
    ~
elif:
    ~
else:
    ~
```

sungchul2 replied:
Done

vinnamkim commented on nncf/experimental/onnx/datasets/imagenet_dataset.py:
```diff
@@ -47,18 +47,32 @@ def __len__(self) -> int:
-    assert len(input_shape) == 4
-
-    return input_shape
+        if len(_input.type.tensor_type.shape.dim) == 4:
+            dim = _input.type.tensor_type.shape.dim
+            _input_shape = [int(MessageToDict(d).get("dimValue")) for d in dim]
+            if main_shape is not None:
+                if main_shape == _input_shape:
+                    input_shape = main_shape
+                    input_keys = _input.name
+                    break
+            else:
+                if input_shape is None and input_keys is None:
+                    input_shape = _input_shape
+                    input_keys = _input.name
+                else:
+                    if input_shape[2:] < _input_shape[2:]:
+                        # If _input_shape > input_shape, input_shape is not a real input size, 
+                        # so replace it with _input_shape.
+                        input_shape = _input_shape
+                        input_keys = _input.name
```
First of all, I think it's more natural to ask the user the name of the input tensor rather than the shape of the input tensor.
Secondly, we need to avoid nested if-else statements for the better readability.
```
assert len(model.graph.input) > 0

if main_shape is given:
     for loop
else:
     input_shape = model.graph.input[0].name
     input_keys = model.graph.input[0].name
     for loop
```

sungchul2 replied:
I fixed to ask user either the input tensor's name or shape and to find the proper tensor shape and name. Also, I removed unnecessary nested if-else.

vinnamkim commented:
@sungchul2 
Please rebase this PR on the latest develop branch and push again. This is because of https://github.com/openvinotoolkit/nncf/pull/1248.

sungchul2 commented:
run tensorflow pre-commit tests

sungchul2 commented:
run pytorch pre-commit tests

MaximProshin commented:
@sungchul2 , do you plan to finalize this PR?

sungchul2 commented:
run pytorch pre-commit tests

kshpv commented on tests/experimental/post_training/test_low_opset_model.py:
```diff
@@ -0,0 +1,128 @@
+]
+
+MODELS = [
+    models.resnet18(),
```
I suggest to test the models that are known to failed without this PR. Do they?

sungchul2 replied:
I added models in tests/onnx/quantization/test_classification_models_graph.py.
What additional models should be tested?

kshpv commented on tests/experimental/post_training/test_low_opset_model.py:
```diff
@@ -0,0 +1,128 @@
+    models.densenet121(),
+]
+
+INPUT_SHAPES = [
```
Please use ModelToTest from tests.onnx.quantization.common

sungchul2 replied:
Done

kshpv commented on tests/experimental/post_training/test_low_opset_model.py:
```diff
@@ -0,0 +1,128 @@
+        if not os.path.isdir(onnx_model_dir):
+            os.mkdir(onnx_model_dir)
+        x = torch.randn(input_shape, requires_grad=False)
+        torch.onnx.export(model, x, onnx_model_path, opset_version=8)
```
Could you do more cases not only opset=8, Please add this as a parameter of the test and try with 6,7,8,9

sungchul2 replied:
Done

kshpv commented on tests/experimental/post_training/test_low_opset_model.py:
```diff
@@ -0,0 +1,128 @@
+
+
+@pytest.fixture
+def load_low_opset_model():
```
why do you need @pytest.fixture here? 
Also change the function name. It exports model and then loads it, which is not described in the name

sungchul2 replied:
Done

kshpv commented on tests/experimental/post_training/test_low_opset_model.py:
```diff
@@ -0,0 +1,128 @@
+
+INPUT_SHAPES = [
+    [1, 3, 224, 224],
+    [1, 3, 224, 224],
```
I would like to see the models with different shapes. I mean the models without the batch dim, or the model with several inputs

sungchul2 replied:
It would be better to proceed with a new ticket.

kshpv replied:
See my comment below

kshpv commented on tests/experimental/post_training/test_low_opset_model.py:
```diff
@@ -0,0 +1,128 @@
+"""
```
Please, place thus test in the following directory - tests/onnx

sungchul2 replied:
Done

sungchul2 commented on examples/experimental/onnx/classification/onnx_ptq_classification.py:
```diff
@@ -29,7 +29,8 @@
 def run(onnx_model_path: str, output_model_path: str,
         dataset_path: str, batch_size: int, shuffle: bool, num_init_samples: int,
-        input_shape: Optional[List[int]] = None, ignored_scopes: Optional[List[str]] = None,
+        input_shape: Optional[List[int]] = None, input_keys: Optional[str] = None,
```
It would be better to proceed with a new ticket.

kshpv replied:
The changes is almost the same as in onnx_ptq_classification.py

kshpv commented on examples/experimental/onnx/classification/onnx_ptq_classification.py:
```diff
@@ -38,15 +39,17 @@ def run(onnx_model_path: str, output_model_path: str,
-        raise RuntimeError(
-            f"The number of inputs should be 1(!={len(input_keys)}).")
+    elif input_keys is None:
+        nncf_logger.info(
```
Could you place this logging inside the function infer_input_shape()?

sungchul2 replied:
Done

kshpv commented on tests/onnx/test_low_opset_model.py:
```diff
@@ -0,0 +1,125 @@
+]
+
+MODELS = [
+    models.resnet18(),
```
All of these models have the same input shape and the same input_keys. It doen't make any sense to test all of them. Please use only two of them.
I wanted to see also models having different input_keys and different input_shape. Please, take a look at tests/onnx/data/models. Could you use [Mask R-CNN](https://github.com/onnx/models/tree/main/vision/object_detection_segmentation/mask-rcnn) and [yolov3](https://github.com/onnx/models/tree/main/vision/object_detection_segmentation/yolov3)

kshpv commented on tests/onnx/test_low_opset_model.py:
```diff
@@ -0,0 +1,125 @@
+
+INPUT_SHAPE = [1, 3, 224, 224]
+INPUT_KEY = 'input.1'
+LOW_OPSET_VERSIONS = [6, 7, 8, 9]
```
If opset 6 is not supported, please remove it.

sungchul2 replied:
Done

sungchul2 commented:
run pytorch pre-commit tests

sungchul2 commented:
@kshpv 
I saw models from torchvision just have single input, not multiple inputs, and it is not the purpose of this ticket.
So, I changed them to models from onnx model zoo, which directly have had this issue, such as [densenet](https://github.com/onnx/models/tree/main/vision/classification/densenet-121/model).

And I added [Mask R-CNN](https://github.com/onnx/models/tree/main/vision/object_detection_segmentation/mask-rcnn) and [yolov3](https://github.com/onnx/models/tree/main/vision/object_detection_segmentation/yolov3) for test as you requested.
These models have dynamic input_shape and have been converted with high opset version (10, 12).
So, I separately added a test for them and renamed the test file.

kshpv commented:
@sungchul2 thank you for the PR. It looks very good!

vshampor commented on examples/experimental/onnx/classification/onnx_ptq_classification.py:
```diff
@@ -101,9 +97,11 @@ def run(onnx_model_path: str, output_model_path: str,
+        "--input_shape", help="Model's input shape. e.g. [1, 3, 224, 224].",
+        nargs="+", type=int, default=None)
+    parser.add_argument(
+        "--input_keys", help="Model's input key.", type=str, default=None)
```
What is an "input key"? Explanation is missing. Is "key" a proper word for this, something that has the expected meaning in the ONNX lexicon?

sungchul2 replied:
Because `input_keys` have been used in sample code, `onnx_ptq_classification.py` and `onnx_ptq_segmentation.py`, I just used `input_keys`. But I checked that ONNX models use `name`, not `key`, and I changed every part that uses `input_keys` to `input_name`.

vshampor commented on examples/experimental/onnx/semantic_segmentation/onnx_ptq_segmentation.py:
```diff
@@ -76,8 +77,11 @@ def run(onnx_model_path: str, output_model_path: str, dataset_name: str,
+        "--input_shape", help="Model's input shape. e.g. [1, 3, 768, 960]",
+        nargs="+", type=int, default=None)
+    parser.add_argument(
+        "--input_keys", help="Model's input key.", type=str, default=None)
```
Furthermore, it is not evident in which format this "key", or multiple "keys" should be given in.

sungchul2 replied:
Added an example similar to `input_shape`.

vshampor commented on nncf/experimental/onnx/datasets/common.py:
```diff
@@ -0,0 +1,68 @@
+from onnx import ModelProto
+
+
+def infer_input_shape(model: ModelProto,
```
Why is this under the `datasets` folder? I don't think that this function relates to datasets at all. The function should also have a docstring, as per [style guide](https://github.com/openvinotoolkit/nncf/blob/develop/docs/styleguide/PyGuide.md).

sungchul2 replied:
Took this file out of `datasets` folder and created a docstring.

vshampor commented on tests/onnx/test_vaious_input_type.py:
```diff
@@ -0,0 +1,120 @@
+"""
```
The name of the file contains a typo. 
`vaious` -> `various`. 

Anyway, this file should have been probably better called `test_infer_input_shape.py`, since this is the functionality you are testing here.

sungchul2 replied:
Renamed to `test_infer_input_shape.py`.

vshampor commented on examples/experimental/onnx/classification/onnx_ptq_classification.py:
```diff
@@ -101,9 +97,11 @@ def run(onnx_model_path: str, output_model_path: str,
-    parser.add_argument("--input_shape",
-        help="Model's input shape. e.g. [1, 3, 224, 224]. If it's not given, it automatically infers input shape.",
-                        nargs="+", type=int, default=None)
+    parser.add_argument(
```
You are duplicating code here in the segmentation sample. Extract the common parser part like it is done in the [PT](examples/torch/common/argparser.py) and [TF](examples/tensorflow/common/argparser.py) samples.

sungchul2 replied:
Done

sungchul2 commented:
retest this please

sungchul2 commented:
run onnx pre-commit tests
